src/transport/SessionManager.cpp

```diff
@@ -1057,27 +1240,59 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & partialPack
-                                                                 const Optional<Transport::SecureSession::Type> & type)
+                                                                 const Optional<Transport::SecureSession::Type> & type,
+                                                                 TransportPayloadCapability transportPayloadCapability)
 {
-    SecureSession * found = nullptr;
-
-    mSecureSessions.ForEachSession([&peerNodeId, &type, &found](auto session) {
+    SecureSession * mrpSession = nullptr;
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    SecureSession * tcpSession = nullptr;
+#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT
+
+    mSecureSessions.ForEachSession([&peerNodeId, &type, &mrpSession,
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+                                    &tcpSession,
+#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT
+                                    &transportPayloadCapability](auto session) {
         if (session->IsActiveSession() && session->GetPeer() == peerNodeId &&
             (!type.HasValue() || type.Value() == session->GetSecureSessionType()))
         {
-            //
-            // Select the active session with the most recent activity to return back to the caller.
-            //
-            if ((found == nullptr) || (found->GetLastActivityTime() < session->GetLastActivityTime()))
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+            if ((transportPayloadCapability == TransportPayloadCapability::kMRPOrTCPCompatiblePayload ||
+                 transportPayloadCapability == TransportPayloadCapability::kLargePayload) &&
+                session->GetTCPConnection() != nullptr)
+            {
+                // Set up a TCP transport based session as standby
+                if ((tcpSession == nullptr) || (tcpSession->GetLastActivityTime() < session->GetLastActivityTime()))
+                {
+                    tcpSession = session;
+                }
+            }
+#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT
+
+            if ((mrpSession == nullptr) || (mrpSession->GetLastActivityTime() < session->GetLastActivityTime()))
             {
-                found = session;
+                mrpSession = session;
             }
         }
 
         return Loop::Continue;
     });
 
-    return found != nullptr ? MakeOptional<SessionHandle>(*found) : Optional<SessionHandle>::Missing();
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    if (transportPayloadCapability == TransportPayloadCapability::kLargePayload)
+    {
+        return tcpSession != nullptr ? MakeOptional<SessionHandle>(*tcpSession) : Optional<SessionHandle>::Missing();
+    }
+
+    if (transportPayloadCapability == TransportPayloadCapability::kMRPOrTCPCompatiblePayload)
+    {
+        return mrpSession != nullptr
+            ? MakeOptional<SessionHandle>(*mrpSession)
+            : (tcpSession != nullptr ? MakeOptional<SessionHandle>(*tcpSession) : Optional<SessionHandle>::Missing());
```

Comment:
This is really hard to follow.  Maybe just write it out with `if` and multiple return statements?

Reply:
Separated into multiple ifs and returns.